### PR TITLE
Documenting Cucumber varieties

### PR DESCRIPTION
**Pull request by aslakhellesoy**
@vincent-psarga and I have discussed a new way to group Cucumber versions in *varieties*. This page explains what varieties are and why we think they are useful. You can see a [preview on Netlify](https://5daf04f8a7304c0008baf5ef--cucumber.netlify.com/docs/cucumber/varieties/)

The *Arola* variety describes the commonalities between the last stable releases of Cucumber.js, Cucumber-JVM and Cucumber-Ruby. 

The *Babylon* variety describes what we are currently working towards.

I don't know what to put in *Bella* (after Babylon). I think the approach we should take is to wait and see where things evolve, and then nail down what we think all Cucumbers should have.

I have not included SpecFlow, Behave or Behat, but I'd love to include them in the table if someone updates the table! 

If we can agree on this variety approach, we should amend the release process for Cucumbers:
* Provide a link to the variety if a release is compatible with it
* Update the varieties page with links to the new release

My biggest hope with Cucumber varieties is that implementations will slowly converge towards a common set of characteristics, and also provide a yard stick for contributors so they know what's high priority. /cc @ciaranmcnulty @jenisys

### CODE REVIEW

File: content/docs/cucumber/varieties.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+* Define stretch goals for upcoming Cucumber implementations
+* Allow tool vendors to support particular varieties with more ease
+* Documentation becomes easier to read and write as it applies to all releases in the variety
+
+Varieties are named after Wikipedia's [list of Cucumber varieties](https://en.wikipedia.org/wiki/List_of_cucumber_varieties). Higher alphabetical order means more recent.
+
+# Babylon
+
+Characteristics               | [Cucumber-JVM 5.0.0-RC2] | [Cucumber.js master] | [Cucumber-Ruby 4.0.0.rc.2]
```
Comment (charlierudolph) [inline content/docs/cucumber/varieties.md:22]: Why is cucumber-js master here instead of a version?

**Comment by mattwynne** [comment]
For me, the word "variety" was confusing. It implies nothing about age / maturity the way that another word like "release" might. Maybe that's deliberate but my first impression was that it was a way to refer to which programming language / platform it worked on. I tend to use "flavour" for that, but it rang the same bell in my mind.

Other than that, I love this model and the philosophy of gently guiding everyone towards convergence.

File: content/docs/cucumber/varieties.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+If not, point to the latest RC release. If no RC release, point to master. 
+-->
+[Cucumber-JVM 5.0.0-RC2]: https://github.com/cucumber/cucumber-jvm/blob/master/CHANGELOG.md#500-rc2-YYYY-mm-dd
+[Cucumber-JVM 4.8.0]: https://github.com/cucumber/cucumber-jvm/blob/master/CHANGELOG.md#480-2019-10-19
+
+[Cucumber-Ruby 4.0.0.rc.2]: https://github.com/cucumber/cucumber-ruby/blob/master/CHANGELOG.md#400rc2-YYYY-mm-dd
+[Cucumber-Ruby 3.1.2]: https://github.com/cucumber/cucumber-ruby/blob/master/CHANGELOG.md#312-2018-07-13
+
+[Cucumber.js master]: https://github.com/cucumber/cucumber-js
+[Cucumber.js master]: https://github.com/cucumber/cucumber-js
```
Comment (olleolleolle) [inline content/docs/cucumber/varieties.md]:
```suggestion
[Cucumber.js 6.0.0]: https://github.com/cucumber/cucumber-js/blob/v6.0.2/CHANGELOG.md#602-2019-10-07
```

Something like this would allow us to point at the 6.0.x series. The name of the link _could_ change, to reflect _6.0.x_-ness.

File: content/docs/cucumber/varieties.md (hunk 1 above)
Comment (aslakhellesoy) [reply content/docs/cucumber/varieties.md:22]: For the most recent variety definition (Babylon) my idea is that the Cucumber version in the header should be one of the following:

* A final (non-RC) release, if it satisfies all of the characteristics
* An RC release, if it satisfies one or more charactersitics (with the assumption that the final release will support them all)
* If none of the above, use master

Does that make sense @charlierudolph? If so, I can add that to the description.

**Comment by aslakhellesoy** [comment]
> For me, the word "variety" was confusing. It implies nothing about age / maturity the way that another word like "release" might.

@mattwynne the maturity level is not in the variety name, but in the lexical/alphabetical order as well as the Cucumber versions in the header.

I'm reluctant to use the word "release" as it is too easy to confuse with the semver versions.

File: content/docs/cucumber/varieties.md (hunk 2 above)
Comment (aslakhellesoy) [reply content/docs/cucumber/varieties.md]: @olleolleolle thanks for fixing that typo. I think we should be explicit with major.minor.version because it's entirely possible that a Cucumber bumps the major version for a *different* reason that the characteristics defined by the variety.

For example, Cucumber-JVM 5.0.0 may be a Babylon, but we may make a 6.0.0 release that does nothing except drop support for Java 8, and so 6.0.0 is still a Babylon.

**Comment by charlierudolph** [comment]
Thoughts on naming these something like `convergent milestone` or `convergence milestone`?

**Comment by aslakhellesoy** [comment]
> Thoughts on naming these something like convergent milestone or convergence milestone

I like that! How about just "milestone"?

**Comment by gasparnagy** [comment]
I like the concept. I was thinking if it would be helpful to have an even earlier milestone/variety that would represent the versions where there are not even cucumber expressions used. (many of the users and tools are still in this level). Ah we could also call these levels, but I am fine with the other names too.

**Comment by oscaralanpierce** [comment]
I'm not totally clear on what would differentiate one variety from the other, but off the top of my head, I think this might make things more confusing for users and also might even have the opposite effect from what you're imagining @aslakhellesoy. If I understand correctly, there could be multiple varieties at each given maturity level. Would these be differentiated by language, or by some more nebulous concept of characteristics? If the latter, I'd find it overwhelming as a user to figure out which one I'd need to use, and since a lot of people are already confused or intimidated by Cucumber, I think that's something we should take into account.

Of course, if I've misunderstood what you're proposing, feel free to ignore :)

**Comment by vincent-psarga** [comment]
> > Thoughts on naming these something like convergent milestone or convergence milestone
> 
> I like that! How about just "milestone"?

I'm not convinced by simply "milestone", for the same reason you're reluctant to use __release__ :)

I think I'd be easier that when we talk about a milestone, we refer to an implementation of Cucumber in a language and when we talk about a __convergent milestone__ (or __convergence milstone__), we talk about all the implementations.

That's where I think the word __variety__ is interesting as it may be less confusing ("may be" because I don't know how people who are not involved in this discussion will understand this notion)

(and sorry for the close/reopen, misclick :/ )

File: content/docs/cucumber/varieties.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+* [cucumber-html-formatter] 2.x
+* [cucumber-json-formatter] 2.x
+
+# Arola
+
+Characteristics               | [Cucumber-JVM 4.8.0] | [Cucumber.js 6.0.2] | [Cucumber-Ruby 3.1.2]
+------------------------------|----------------------|---------------------|---------------------------
+[gherkin] 5.x                 |                   ✅ |                  ✅ |            ✅ 
+[tag-expressions] 2.0.x       |                   ✅ |                  ✅ |            ❌ (`~> 1.1.0`)
```
Comment (luke-hill) [inline content/docs/cucumber/varieties.md:40]: given what you have done here with the version available for cucumber ruby (Being lower than the required one). And above I'm guessing not implemented. Should we maybe have some indication it's not implemented yet?

**Comment by mlvandijk** [comment]
While I **really** like the idea of:
1. having an overview which features (if that's the right term) are available in which version of each implementation, and
2. converging to having the *same* functionalities available acrosss implementations,
I personally find the term "varieties" (or "milestones") confusing.

Would it be an idea to just list "features" / functionalities, like Gherkin version, Cucumber expressions, and then list per implementation as of which version that feature is available? (or X if not available). I think that would be hard enough to understand, without adding a new term (variety/milestone/whatever).

File: content/docs/community/blog-posts.md [hunk 4]
```diff
@@ -14,8 +14,6 @@ See also [Related Tools](/docs/tools/related-tools).
...
-| [Getting Started With Cucumber Pro](https://jayson.codes/and/blogs/2018/08/getting-started-with-cucumber-pro/)                                                                                                                                                         | Getting up and running with Cucumber Pro and some handy tips by Jayson Smith                                                                          | August 9, 2018      |
-| [Closing A Two Year Old Cucumber Ruby Issue](https://jayson.codes/and/blogs/2018/08/cucumber-closing-1021/)                                                                                                                                                            | Jayson Smith goes over closing a two year old issue in the Ruby Cucumber project around code style                                                    | August 5, 2018      |
```
Comment (mpkorstanje) [inline content/docs/community/blog-posts.md]: Can we split the boy scouting into a separate MR? Then we can get those improvements through without being held up by the discussion.

Comment (mpkorstanje) [reply content/docs/community/blog-posts.md]: Done.

File: content/docs/cucumber/varieties.md (hunk 1 above)
Comment (mpkorstanje) [inline content/docs/cucumber/varieties.md:22]: Cucumber JVM 5.0.0-RC2 hasn't been released yet. We should only include released versions here. Otherwise this becomes a speculative document rather then a reference.

**Comment by aslakhellesoy** [comment]
This PR was supposed to target the branch for #400 rather than master. That’s why changes from 400 shows up here too.

File: content/docs/cucumber/varieties.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+weight: 3
+---
+
+Every [Cucumber implementation](/docs/installation) has its own codebase, release cycle and [semver](https://semver.org/) version that is independent from other Cucumber implementations.
+
+The Cucumber team also maintains a list of *varieties* which defines essential characteristics of various Cucumber versions. A Cucumber release may satisfy zero or more characteristics of a variety. If it satisfies all of them it is a member of that variety.
+
+Marking Cucumber releases as part of a variety serves several purposes:
+
+* Define stretch goals for upcoming Cucumber implementations
```
Comment (mpkorstanje) [inline content/docs/cucumber/varieties.md:14]: I'd leave this out. The complexity of the code base, deprecation strategy and the feature itself tends to influence when it can be released more so then our mere desire to include it.

Comment (mpkorstanje) [reply content/docs/cucumber/varieties.md:14]: Given that this is `doc.cucumber.io` it would also be better for this document to be factual rather then speculative.

**Comment by aslakhellesoy** [comment]
Having discussed this with @aurelien-reeves we've decided to abandon this. It would be too hard to maintain.
